synapse/http/server.py

```diff
@@ -870,10 +871,20 @@ def set_cors_headers(request: Request) -> None:
     request.setHeader(
         b"Access-Control-Allow-Methods", b"GET, HEAD, POST, PUT, DELETE, OPTIONS"
     )
-    request.setHeader(
-        b"Access-Control-Allow-Headers",
-        b"X-Requested-With, Content-Type, Authorization, Date",
-    )
+    if getattr(request, "experimental_cors_msc3886", False):
```

Comment:
also wondering why we need `getattr` here — just generally cautious about it because it evades typechecking etc

Reply:
The original rationale for this one was being cautious around subclasses of `SynapseRequest` such as `XForwardedForRequest` and mocks.

I will try with a direct property and see what breaks.

Reply:
subclasses should inherit the attribute. We shouldn't be afraid of breaking mocks, but even mocks usually will use `spec=SynapseRequest` so they'll inherit some slots for the fields I believe.

Reply:
I've updated the PR to make the typing of `SynapseRequest` more explicit which has bubbled up to a number of other places in the codebase.

If this is too invasive I can change it to use `isinstance` so that the types can remain as `Request` not `SynapseRequest`.